### PR TITLE
refactor: enable withdraw all stake endpoint

### PR DESCRIPTION
**Pull request by swarmHaseeb**
### Checklist

- [x] I have read the [coding guide](https://github.com/ethersphere/bee/blob/master/CODING.md).
- [x] My change requires a documentation update, and I have done it.
- [x] I have added tests to cover my changes.
- [x] I have filled out the description and linked the related issues.

### Description

Adds an endpoint to delete all existing stake.
### Screenshots (if appropriate):

- - -
This change is [Reviewable](https://reviewable.io/reviews/ethersphere/bee/3662)

### CODE REVIEW

File: pkg/api/staking.go [hunk 1]
```diff
@@ -90,3 +95,23 @@ func (s *Service) getStakedAmountHandler(w http.ResponseWriter, r *http.Request)
...
+
+	txHash, err := s.stakingContract.WithdrawAllStake(r.Context())
+	if err != nil {
+		if errors.Is(err, staking.ErrInsufficientStake) {
+			logger.Debug("insufficient stake", "overlayAddr", s.overlay, "error", err)
+			logger.Error(nil, "insufficient stake")
+			jsonhttp.BadRequest(w, "insufficient stake to withdraw")
+			return
+		}
+		logger.Debug("delete stake failed", "error", err)
```
Comment (istae) [inline pkg/api/staking.go:110]: replace the word delete

File: pkg/storageincentives/staking/contract_test.go [hunk 2]
```diff
@@ -635,3 +635,404 @@ func TestGetStake(t *testing.T) {
 		}
 	})
 }
+
+func TestDeleteStake(t *testing.T) {
```
Comment (istae) [inline pkg/storageincentives/staking/contract_test.go:639]: withdraw

File: pkg/storageincentives/staking/contract.go [hunk 3]
```diff
@@ -224,3 +228,74 @@ func (c *contract) getBalance(ctx context.Context) (*big.Int, error) {
...
+	receipt, err := c.withdrawFromStake(ctx, stakedAmount)
+	if receipt != nil {
+		txHash = receipt.TxHash
+	}
+	return
+}
+
+func (c *contract) withdrawFromStake(ctx context.Context, stakedAmount *big.Int) (*types.Receipt, error) {
+	var overlayAddr [32]byte
+	copy(overlayAddr[:], c.overlay.Bytes())
```
Comment (mrekucci) [inline pkg/storageincentives/staking/contract.go:266]: This is not needed, you can use `c.overlay.Clone().Bytes()` instead.

File: pkg/storageincentives/staking/contract.go [hunk 4]
```diff
@@ -224,3 +228,74 @@ func (c *contract) getBalance(ctx context.Context) (*big.Int, error) {
...
+	if err != nil {
+		return false, err
+	}
+
+	results, err := c.stakingContractABI.Unpack("paused", result)
+	if err != nil {
+		return false, err
+	}
+
+	return results[0].(bool), nil
```
Comment (mrekucci) [inline pkg/storageincentives/staking/contract.go:300]: It is not guaranteed that there will be at least one element in the result.

File: pkg/storageincentives/staking/contract.go (hunk 3 above)
Comment (istae) [reply pkg/storageincentives/staking/contract.go:266]: I think the abi packing needs a [32]byte and not a byte slice

File: openapi/SwarmDebug.yaml [hunk 5]
```diff
@@ -1056,6 +1056,23 @@ paths:
           $ref: "SwarmCommon.yaml#/components/responses/500"
         default:
           description: Default response
+    delete:
+      summary: Withdraw all staked amount.
+      description: Be aware, this endpoint creates an on-chain transactions and transfers BZZ from the node's Ethereum account and hence directly manipulates the wallet balance.
```
Comment (notanatol) [inline openapi/SwarmDebug.yaml:1061]: minor: we can either drop the `an` or use `transaction` (singular)

**Review by mrekucci — APPROVED** [review]

File: pkg/storageincentives/staking/contract.go (hunk 3 above)
Comment (swarmHaseeb) [reply pkg/storageincentives/staking/contract.go:276]: nope, it needs byte slice

File: pkg/api/staking.go [hunk 6]
```diff
@@ -90,3 +95,23 @@ func (s *Service) getStakedAmountHandler(w http.ResponseWriter, r *http.Request)
 
 	jsonhttp.OK(w, getStakeResponse{StakedAmount: bigint.Wrap(stakedAmount)})
 }
+
+func (s *Service) withdrawAllStakeHandler(w http.ResponseWriter, r *http.Request) {
+	logger := s.logger.WithName("withdraw_all_stake").Build()
+
+	txHash, err := s.stakingContract.WithdrawAllStake(r.Context())
+	if err != nil {
+		if errors.Is(err, staking.ErrInsufficientStake) {
```
[the remark this answers is not in the document]
Comment (swarmHaseeb) [reply pkg/api/staking.go:104]: nope, this edge case deals with the fact that we did not have any stake to begin with

**Review by istae — APPROVED** [review]

**Pull request merged**